src/edu/csus/ecs/pc2/clics/API202306/RunService.java

```diff
@@ -0,0 +1,145 @@
+    public Response getRuns(@Context SecurityContext sc, @PathParam("contestId") String contestId) {
+        
+        // check contest id
+        if(contestId.equals(model.getContestIdentifier()) == false) {
+            return Response.status(Response.Status.NOT_FOUND).build();        
+        }
+        
+        long freezeTime = Utilities.getFreezeTime(model);
+        
+        ArrayList<CLICSTestCase> tclist = new ArrayList<CLICSTestCase>();
+        
+        for (Run run: model.getRuns()) {
+            // If not admin or judge, can not see runs after freeze time
+            if (!sc.isUserInRole(WebServer.WEBAPI_ROLE_ADMIN) && !sc.isUserInRole(WebServer.WEBAPI_ROLE_JUDGE)) {
+                // if run is after scoreboard freeze, do not return info for it
+                if (run.getElapsedMS() / 1000 > freezeTime) {
+                    continue;
+                }
+            }
+            if(run.isJudged() && !run.getJudgementRecord().isPreliminaryJudgement()) {
+                for(RunTestCase testCase: run.getRunTestCases()) {
+                    tclist.add(new CLICSTestCase(model, testCase));
+                }
+            }
+        }
+        try {
+            ObjectMapper mapper = JSONUtilities.getObjectMapper();
+            String json = mapper.writeValueAsString(tclist);
+            return Response.ok(json, MediaType.APPLICATION_JSON).build();
+        } catch (Exception e) {
+            return Response.status(Status.INTERNAL_SERVER_ERROR).entity("Error creating JSON for Runs testcases " + e.getMessage()).build();
+        }
+    }
+
+    /**
+     * Returns a representation of the specified test case in the specified contest in JSON format. The returned value is compliant with 2023-06 API.
+     * 
+     * @param sc User's info
+     * @param contestId The contest
+     * @param runId The run of interest
+     * @return response
+     */
+    @GET
+    @Produces(MediaType.APPLICATION_JSON)
+    @Path("{runId}/")
+    public Response getRun(@Context SecurityContext sc, @PathParam("contestId") String contestId, @PathParam("runId") String runId) {
+
+        // check contest id
+        if(contestId.equals(model.getContestIdentifier()) == true) {
+            long freezeTime = Utilities.getFreezeTime(model);
+    
+            for (Run run: model.getRuns()) {
+                // If not admin or judge, can not see runs after freeze time
+                if (!sc.isUserInRole(WebServer.WEBAPI_ROLE_ADMIN) && !sc.isUserInRole(WebServer.WEBAPI_ROLE_JUDGE)) {
+                    // if run is after scoreboard freeze, do not return info for it
+                    if (run.getElapsedMS() / 1000 > freezeTime) {
+                        continue;
+                    }
+                }
+                if(run.isJudged() && !run.getJudgementRecord().isPreliminaryJudgement()) {
+                    for(RunTestCase testCase: run.getRunTestCases()) {
+                        if (testCase.getElementId().toString().equals(runId)) {
```

Comment:
It bothers me greatly that we are sending out PC2 `ElementId` values as run (testcase) ids.  Those don't seem like good values to be exposing externally.  But I don't know what to do about it...

Reply:
Yeah, me neither.  I guess we could something like *cough* Kattis does, and have a monotonically increasing unique number in the system they use for ID's, and every time we need one, set it to that.  But I don't see the advantage of that over a random ID which is what we do now (sort of).